In [1]:
import pandas as pd
import numpy as np

In [2]:
ted_data = pd.read_csv("ted_main.csv")
youtube_data = pd.read_csv("ted_youtube.csv")

In [3]:
ted_data.drop(['description','film_date','published_date','url'],axis=1,inplace=True)

In [4]:
ted_data=ted_data.rename({'comments':'ted_nof_comments','duration':'ted_duration','event':'ted_event','languages':'ted_avlb_lang','published_dt':
                'ted_published_dt','ratings':'ted_ratings','related_talks':'ted_talks_related','tags':'ted_tags','title':'ted_title','views':'ted_views'},axis=1)

In [5]:
youtube_data.drop(['Unnamed: 0','fulltitle','webpage_url','uploader_id','description','fulltitle_compare','id'],axis=1,inplace=True)

In [6]:
youtube_data = youtube_data.rename({'view_count':'youtube_views','duration':'youtube_duration','categories':'youtube_categories','tags':'youtube_tags',
                    'dislike_count':'youtube_dislikes','like_count':'youtube_likes','average_rating':'youtube_avg_ratings','title':'youtube_title'},axis=1)

In [7]:
speaker_occupation_na = ['Writer, educator, CEO','Entrepreneur, Politician','Technologist, CEO','Musician','Artist','Photographer, Artist, Activist']
ted_data.loc[ted_data.speaker_occupation.isna(),'speaker_occupation'] = speaker_occupation_na

In [8]:
youtube_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2492 entries, 0 to 2491
Data columns (total 8 columns):
youtube_views          2492 non-null int64
youtube_title          2492 non-null object
youtube_tags           2492 non-null object
youtube_categories     2492 non-null object
youtube_duration       2492 non-null int64
youtube_dislikes       2482 non-null float64
youtube_likes          2482 non-null float64
youtube_avg_ratings    2492 non-null float64
dtypes: float64(3), int64(2), object(3)
memory usage: 155.8+ KB


In [9]:
youtube_data['youtube_likes']=youtube_data['youtube_likes'].fillna(0)
youtube_data['youtube_dislikes']=youtube_data['youtube_dislikes'].fillna(0)

In [10]:
youtube_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2492 entries, 0 to 2491
Data columns (total 8 columns):
youtube_views          2492 non-null int64
youtube_title          2492 non-null object
youtube_tags           2492 non-null object
youtube_categories     2492 non-null object
youtube_duration       2492 non-null int64
youtube_dislikes       2492 non-null float64
youtube_likes          2492 non-null float64
youtube_avg_ratings    2492 non-null float64
dtypes: float64(3), int64(2), object(3)
memory usage: 155.8+ KB


In [11]:
youtube_data['melt1'] = youtube_data.youtube_title.str.split(':')

youtube_data['talk1']=youtube_data['melt1'].str.get(1)
youtube_data['author1']=youtube_data['melt1'].str.get(0)

In [12]:
youtube_data['melt2'] = youtube_data.youtube_title.str.split('|')

youtube_data['talk2']=youtube_data['melt2'].str.get(0)
youtube_data['author2']=youtube_data['melt2'].str.get(1)

In [13]:
youtube_data

youtube_views                                      youtube_title  \
0             20695  Inge Missmahl brings peace to the minds of Afg...   
1            129968  Edith Widder: The weird and wonderful world of...   
2             46844     Jacqueline Novogratz on an escape from poverty   
3            281435            Math class needs a makeover | Dan Meyer   
4            192313              Craig Venter unveils "synthetic life"   
5             50511              Danielle de Niese: A flirtatious aria   
6             50557  Philip Howard: Four ways to fix a broken legal...   
7            981094  Why you should define your fears instead of yo...   
8             35683    Dennis vanEngelsdorp: Where have the bees gone?   
9           1751099   Close-up card magic with a twist | Lennart Green   
10           350079            Joshua Klein: The intelligence of crows   
11          1861625           The psychology of evil | Philip Zimbardo   
12           627112              How to make hard choices | Ruth Chang   
13            61947  Dan Barasch: A park underneath the hustle and ...   
14           137256  Jill Tarter: Why the search for alien intellig...   
15            36779              Liz Diller: A giant bubble for debate   
16            42353  Seth Shostak: ET is (probably) out there — get...   
17           112552                   Joseph Pine: What consumers want   
18           109921  Jinha Lee: Reach into the computer and grab a ...   
19            94321        Clay Shirky: Institutions vs. collaboration   
20           729312  One woman, eight hilarious characters | Sarah ...   
21           287048               How to find a wonderful idea | OK Go   
22           827676  Hans Rosling: Global population growth, box by...   
23            61354   Lian Pin Koh: A drone's-eye view of conservation   
24            33184  Jessica Green: Are we filtering the wrong micr...   
25           110647  Lee Mokobe: A powerful poem about what it feel...   
26            44115  Roger Stein: A bold new way to fund drug research   
27           234990    Why our screens make us less happy | Adam Alter   
28          1289448             Underwater astonishments | David Gallo   
29            77605  How radio telescopes show us unseen galaxies |...   
...             ...                                                ...   
2462         225847                Stuart Brown: Play is more than fun   
2463          30407  Boghuma Kabisen Titanji: Ethical riddles in HI...   
2464         103549              Bruno Maisonnier: Dance, tiny robots!   
2465          50790  Pearl Arredondo: My story, from gangland daugh...   
2466        1084846  Incredible high school musicians from Venezuel...   
2467         477762  On tennis, love and motherhood | Serena Willia...   
2468       13838924     Do schools kill creativity? | Sir Ken Robinson   
2469         150652      Aaron Huey: America's native prisoners of war   
2470         145161  Lessons from the longest study on human develo...   
2471          16645  Rob Harmon: How the market can keep streams fl...   
2472          61998         Emily Pilloton: Teaching design for change   
2473          94867              Ed Boyden: A light switch for neurons   
2474          39603          Alan Siegel: Let's simplify legal jargon!   
2475         349590                            Ze Frank's web playroom   
2476          95681  Seth Priebatsch: Building the game layer on to...   
2477          63575  George Ayittey: Cheetahs vs. Hippos for Africa...   
2478         173724  Marcin Jakubowski: Open-sourced blueprints for...   
2479         223198  Susan Savage-Rumbaugh: The gentle genius of bo...   
2480         112813  Rajesh Rao: Computing a Rosetta Stone for the ...   
2481          81385    Paul Debevec animates a photo-real digital face   
2482         127911  George Monbiot: For more wonder, rewild the world   
2483          96338  How (and why) Russia hacked the US election | ...   
2484      

In [14]:
youtube_data['talk3']= youtube_data.youtube_title[(youtube_data.author2.isna()) & (youtube_data.talk1.isna())]

In [15]:
youtube_data['talk3']=youtube_data['talk3'].fillna(youtube_data['author1'].str.cat(youtube_data['talk1'],sep=' '))
youtube_data['talk3']=youtube_data['talk3'].fillna(youtube_data['author2'].str.cat(youtube_data['talk2'],sep=' '))

In [16]:
youtube_data

youtube_views                                      youtube_title  \
0             20695  Inge Missmahl brings peace to the minds of Afg...   
1            129968  Edith Widder: The weird and wonderful world of...   
2             46844     Jacqueline Novogratz on an escape from poverty   
3            281435            Math class needs a makeover | Dan Meyer   
4            192313              Craig Venter unveils "synthetic life"   
5             50511              Danielle de Niese: A flirtatious aria   
6             50557  Philip Howard: Four ways to fix a broken legal...   
7            981094  Why you should define your fears instead of yo...   
8             35683    Dennis vanEngelsdorp: Where have the bees gone?   
9           1751099   Close-up card magic with a twist | Lennart Green   
10           350079            Joshua Klein: The intelligence of crows   
11          1861625           The psychology of evil | Philip Zimbardo   
12           627112              How to make hard choices | Ruth Chang   
13            61947  Dan Barasch: A park underneath the hustle and ...   
14           137256  Jill Tarter: Why the search for alien intellig...   
15            36779              Liz Diller: A giant bubble for debate   
16            42353  Seth Shostak: ET is (probably) out there — get...   
17           112552                   Joseph Pine: What consumers want   
18           109921  Jinha Lee: Reach into the computer and grab a ...   
19            94321        Clay Shirky: Institutions vs. collaboration   
20           729312  One woman, eight hilarious characters | Sarah ...   
21           287048               How to find a wonderful idea | OK Go   
22           827676  Hans Rosling: Global population growth, box by...   
23            61354   Lian Pin Koh: A drone's-eye view of conservation   
24            33184  Jessica Green: Are we filtering the wrong micr...   
25           110647  Lee Mokobe: A powerful poem about what it feel...   
26            44115  Roger Stein: A bold new way to fund drug research   
27           234990    Why our screens make us less happy | Adam Alter   
28          1289448             Underwater astonishments | David Gallo   
29            77605  How radio telescopes show us unseen galaxies |...   
...             ...                                                ...   
2462         225847                Stuart Brown: Play is more than fun   
2463          30407  Boghuma Kabisen Titanji: Ethical riddles in HI...   
2464         103549              Bruno Maisonnier: Dance, tiny robots!   
2465          50790  Pearl Arredondo: My story, from gangland daugh...   
2466        1084846  Incredible high school musicians from Venezuel...   
2467         477762  On tennis, love and motherhood | Serena Willia...   
2468       13838924     Do schools kill creativity? | Sir Ken Robinson   
2469         150652      Aaron Huey: America's native prisoners of war   
2470         145161  Lessons from the longest study on human develo...   
2471          16645  Rob Harmon: How the market can keep streams fl...   
2472          61998         Emily Pilloton: Teaching design for change   
2473          94867              Ed Boyden: A light switch for neurons   
2474          39603          Alan Siegel: Let's simplify legal jargon!   
2475         349590                            Ze Frank's web playroom   
2476          95681  Seth Priebatsch: Building the game layer on to...   
2477          63575  George Ayittey: Cheetahs vs. Hippos for Africa...   
2478         173724  Marcin Jakubowski: Open-sourced blueprints for...   
2479         223198  Susan Savage-Rumbaugh: The gentle genius of bo...   
2480         112813  Rajesh Rao: Computing a Rosetta Stone for the ...   
2481          81385    Paul Debevec animates a photo-real digital face   
2482         127911  George Monbiot: For more wonder, rewild the world   
2483          96338  How (and why) Russia hacked the US election | ...   
2484      

removing white space and changing all cases to loweer without spl char

In [17]:
youtube_data['talk3']=youtube_data['talk3'].str.replace('[^A-Za-z0-9]+','')
youtube_data['talk3']=youtube_data['talk3'].str.strip()
youtube_data['talk3']=youtube_data['talk3'].str.lower()

as ted name contains no : we are removing

In [18]:
ted_data['name']=ted_data['name'].str.replace(':','')
ted_data['name']=ted_data['name'].str.replace('[^A-Za-z0-9]+','')
ted_data['name']=ted_data['name'].str.strip()
ted_data['name']=ted_data['name'].str.lower()

in you tube data we have '(with english subtitles) in descr so removing

In [19]:
youtube_data.talk3[youtube_data['talk3'].str.contains('withenglish')]

536     carinamorillotounderstandautismdontlookawaywit...
661     karinagalperinshouldwesimplifyspellingwithengl...
687     jorgedrexlerpoetrymusicandidentitywithenglishs...
932     jorgeramoswhyjournalistshaveanobligationtochal...
996     emilianosalinasacivilresponsetoviolencewitheng...
1206    tomassaracenowouldyouliveinafloatingcityinthes...
1596    gabrielagonzlezhowligodiscoveredgravitationalw...
1859    rominalibsterthepowerofherdimmunitywithenglish...
2021    gonzalovilariohowargentinasblindsoccerteambeca...
2141    eduardosenzdecabeznmathisforeverwithenglishsub...
2351    cristinadomenechpoetrythatfreesthesoulwithengl...
Name: talk3, dtype: object

In [20]:
youtube_data['talk3']=youtube_data['talk3'].str.replace('withenglishsubtitles','')

In [21]:
pd.merge(ted_data,youtube_data,left_on='name',right_on='talk3')

ted_nof_comments  ted_duration       ted_event     film_dt  \
0                  265           977         TED2006   2/25/2006   
1                  124          1286         TED2006   2/24/2006   
2                  200          1116         TED2006   2/26/2006   
3                  672          1305         TED2006    2/2/2006   
4                  919           992         TED2006   2/24/2006   
5                  900          1262         TED2006   2/25/2006   
6                  325          1481         TED2006   2/23/2006   
7                  305          1445         TED2004   2/26/2004   
8                  185          1136         TED2004   2/24/2004   
9                  556          1407         TED2006   2/24/2006   
10                 184          1140  TEDGlobal 2005   7/14/2005   
11                 507          1316  TEDGlobal 2005    7/7/2005   
12                 355          1050         TED2004   2/26/2004   
13                 991          1177  TEDGlobal 2005   7/15/2005   
14                 612          1365  TEDGlobal 2005   7/14/2005   
15                 494           805         TED2006   2/23/2006   
16                 233          1376         TED2005   2/24/2005   
17                 149          1200         TED2005   2/24/2005   
18                  95           247   TEDSalon 2006  11/30/2006   
19                 499          1001         TED2005    2/2/2005   
20                  91          1141  TEDGlobal 2005   7/14/2005   
21                 111          1385         TED2005   2/25/2005   
22                 161          1645         TED2002    3/1/2002   
23                 220          1021         TED2003   2/27/2003   
24                 114          1211         TED2005    2/2/2005   
25                  49          1340         TED2003    2/2/2003   
26                  86           930         TED2004    2/2/2004   
27                 107          1054  TEDGlobal 2005    7/7/2005   
28                  40           276   TEDSalon 2006  12/14/2006   
29                  27           850  TEDGlobal 2005   7/14/2005   
...                ...           ...             ...         ...   
1908                81           736         TED2017   4/24/2017   
1909                13           685    TEDxStanford   4/24/2016   
1910                12           530         TED2017   4/24/2017   
1911                20           747    TEDxCrenshaw   10/8/2016   
1912                 9           310         TED2017   4/24/2017   
1913                18           791      TEDxSydney   5/24/2016   
1914                39           817         TED2017   4/24/2017   
1915                17           865    TEDxStanford   4/23/2017   
1916                 3           717     TEDxHamburg    6/8/2016   
1917                45          1156      TEDxExeter   4/24/2015   
1918                12           738         TED2017   4/24/2017   
1919                18           698       TEDxGhent   6/18/2016   
1920                25           993         TED2017   4/24/2017   
1921                 9           663         TED2017   4/24/2017   
1922                 2           725       TEDxSkoll    4/7/2017   
1923                 8           723  TEDGlobal 2017   8/27/2017   
1924                21           865    TEDxMileHigh    7/8/2017   
1925                39           738         TED2017   4/24/2017   
1926                 7           808      TEDxZurich  11/11/2016   
1927                10           934         TED2017   8/27/2017   
1928                33           722         TED@UPS   7/20/2017   
1929                12           957  TEDxManchester   2/19/2017   
1930                 3           795  TEDGlobal 2017   8/27/2017   
1931                 7           894         TED2017   4/24/2017   
1932                10           810  TEDGlobal 2017   8/27/2017   
1933                17           476         TED2017    6/6/2017   
1934                 6           290         TED2017   4/24/20

soem descriptions differ by few workds so taking slice of text to match

In [22]:
ted_data['name_slice']= ted_data['name'].str.slice(start=0,stop=25)
youtube_data['talk_slice']=youtube_data['talk3'].str.slice(start=0,stop=25)

In [23]:
table_merge1 = pd.merge(ted_data,youtube_data,left_on='name_slice',right_on='talk_slice')
table_merge1

ted_nof_comments  ted_duration       ted_event     film_dt  \
0                  265           977         TED2006   2/25/2006   
1                  124          1286         TED2006   2/24/2006   
2                  200          1116         TED2006   2/26/2006   
3                  672          1305         TED2006    2/2/2006   
4                  919           992         TED2006   2/24/2006   
5                  900          1262         TED2006   2/25/2006   
6                   99          1057         TED2006   2/23/2006   
7                   43          1000         EG 2007  12/12/2007   
8                  325          1481         TED2006   2/23/2006   
9                  305          1445         TED2004   2/26/2004   
10                 185          1136         TED2004   2/24/2004   
11                 556          1407         TED2006   2/24/2006   
12                 184          1140  TEDGlobal 2005   7/14/2005   
13                 507          1316  TEDGlobal 2005    7/7/2005   
14                  95          1275         TED2004   2/27/2004   
15                 355          1050         TED2004   2/26/2004   
16                 991          1177  TEDGlobal 2005   7/15/2005   
17                 612          1365  TEDGlobal 2005   7/14/2005   
18                  58           773  TEDGlobal 2005    7/1/2005   
19                 494           805         TED2006   2/23/2006   
20                 233          1376         TED2005   2/24/2005   
21                 149          1200         TED2005   2/24/2005   
22                  95           247   TEDSalon 2006  11/30/2006   
23                 499          1001         TED2005    2/2/2005   
24                  91          1141  TEDGlobal 2005   7/14/2005   
25                 111          1385         TED2005   2/25/2005   
26                 161          1645         TED2002    3/1/2002   
27                 220          1021         TED2003   2/27/2003   
28                 114          1211         TED2005    2/2/2005   
29                  49          1340         TED2003    2/2/2003   
...                ...           ...             ...         ...   
1963                81           736         TED2017   4/24/2017   
1964                13           685    TEDxStanford   4/24/2016   
1965                12           530         TED2017   4/24/2017   
1966                20           747    TEDxCrenshaw   10/8/2016   
1967                 9           310         TED2017   4/24/2017   
1968                18           791      TEDxSydney   5/24/2016   
1969                39           817         TED2017   4/24/2017   
1970                17           865    TEDxStanford   4/23/2017   
1971                 3           717     TEDxHamburg    6/8/2016   
1972                45          1156      TEDxExeter   4/24/2015   
1973                12           738         TED2017   4/24/2017   
1974                18           698       TEDxGhent   6/18/2016   
1975                25           993         TED2017   4/24/2017   
1976                 9           663         TED2017   4/24/2017   
1977                 2           725       TEDxSkoll    4/7/2017   
1978                 8           723  TEDGlobal 2017   8/27/2017   
1979                21           865    TEDxMileHigh    7/8/2017   
1980                39           738         TED2017   4/24/2017   
1981                 7           808      TEDxZurich  11/11/2016   
1982                10           934         TED2017   8/27/2017   
1983                33           722         TED@UPS   7/20/2017   
1984                12           957  TEDxManchester   2/19/2017   
1985                 3           795  TEDGlobal 2017   8/27/2017   
1986                 7           894         TED2017   4/24/2017   
1987                10           810  TEDGlobal 2017   8/27/2017   
1988                17           476         TED2017    6/6/2017   
1989                 6           290         TED2017   4/24/20

to find the difference between the two dataframes we use outer join

In [24]:
join_outer = pd.merge(ted_data,youtube_data,left_on='name_slice',right_on='talk_slice',indicator=True,how='outer')

In [25]:
join_outer[join_outer['_merge']=='right_only']

ted_nof_comments  ted_duration ted_event film_dt  ted_avlb_lang  \
2553               NaN           NaN       NaN     NaN            NaN   
2554               NaN           NaN       NaN     NaN            NaN   
2555               NaN           NaN       NaN     NaN            NaN   
2556               NaN           NaN       NaN     NaN            NaN   
2557               NaN           NaN       NaN     NaN            NaN   
2558               NaN           NaN       NaN     NaN            NaN   
2559               NaN           NaN       NaN     NaN            NaN   
2560               NaN           NaN       NaN     NaN            NaN   
2561               NaN           NaN       NaN     NaN            NaN   
2562               NaN           NaN       NaN     NaN            NaN   
2563               NaN           NaN       NaN     NaN            NaN   
2564               NaN           NaN       NaN     NaN            NaN   
2565               NaN           NaN       NaN     NaN            NaN   
2566               NaN           NaN       NaN     NaN            NaN   
2567               NaN           NaN       NaN     NaN            NaN   
2568               NaN           NaN       NaN     NaN            NaN   
2569               NaN           NaN       NaN     NaN            NaN   
2570               NaN           NaN       NaN     NaN            NaN   
2571               NaN           NaN       NaN     NaN            NaN   
2572               NaN           NaN       NaN     NaN            NaN   
2573               NaN           NaN       NaN     NaN            NaN   
2574               NaN           NaN       NaN     NaN            NaN   
2575               NaN           NaN       NaN     NaN            NaN   
2576               NaN           NaN       NaN     NaN            NaN   
2577               NaN           NaN       NaN     NaN            NaN   
2578               NaN           NaN       NaN     NaN            NaN   
2579               NaN           NaN       NaN     NaN            NaN   
2580               NaN           NaN       NaN     NaN            NaN   
2581               NaN           NaN       NaN     NaN            NaN   
2582               NaN           NaN       NaN     NaN            NaN   
...                ...           ...       ...     ...            ...   
3025               NaN           NaN       NaN     NaN            NaN   
3026               NaN           NaN       NaN     NaN            NaN   
3027               NaN           NaN       NaN     NaN            NaN   
3028               NaN           NaN       NaN     NaN            NaN   
3029               NaN           NaN       NaN     NaN            NaN   
3030               NaN           NaN       NaN     NaN            NaN   
3031               NaN           NaN       NaN     NaN            NaN   
3032               NaN           NaN       NaN     NaN            NaN   
3033               NaN           NaN       NaN     NaN            NaN   
3034               NaN           NaN       NaN     NaN            NaN   
3035               NaN           NaN       NaN     NaN            NaN   
3036               NaN           NaN       NaN     NaN            NaN   
3037               NaN           NaN       NaN     NaN            NaN   
3038               NaN           NaN       NaN     NaN            NaN   
3039               NaN           NaN       NaN     NaN            NaN   
3040               NaN           NaN       NaN     NaN            NaN   
3041               NaN           NaN       NaN     NaN            NaN   
3042               NaN           NaN       NaN     NaN            NaN   
3043               NaN           NaN       NaN     NaN            NaN   
3044               NaN           NaN       NaN     NaN            NaN   
3045               NaN           NaN       NaN     NaN            NaN   
3046               NaN           NaN       NaN     NaN            NaN   
3047               NaN           NaN       NaN     NaN          

some of the titles differ in ted and youtube data like "Eva Vertes - My dream about future medicine" is the titke in youtube but in ted 
it is "meet the future of camcer research"

We can try to match the remaining fields by author 

In [26]:
youtube_data['split']= youtube_data.talk2[(youtube_data['talk1'].isna())&(youtube_data['author2'].isna())].str.split(' ')

In [27]:
youtube_data['author3']=youtube_data['split'].str.get(0).str.cat(youtube_data['split'].str.get(1),sep=' ')

In [28]:
youtube_data

youtube_views                                      youtube_title  \
0             20695  Inge Missmahl brings peace to the minds of Afg...   
1            129968  Edith Widder: The weird and wonderful world of...   
2             46844     Jacqueline Novogratz on an escape from poverty   
3            281435            Math class needs a makeover | Dan Meyer   
4            192313              Craig Venter unveils "synthetic life"   
5             50511              Danielle de Niese: A flirtatious aria   
6             50557  Philip Howard: Four ways to fix a broken legal...   
7            981094  Why you should define your fears instead of yo...   
8             35683    Dennis vanEngelsdorp: Where have the bees gone?   
9           1751099   Close-up card magic with a twist | Lennart Green   
10           350079            Joshua Klein: The intelligence of crows   
11          1861625           The psychology of evil | Philip Zimbardo   
12           627112              How to make hard choices | Ruth Chang   
13            61947  Dan Barasch: A park underneath the hustle and ...   
14           137256  Jill Tarter: Why the search for alien intellig...   
15            36779              Liz Diller: A giant bubble for debate   
16            42353  Seth Shostak: ET is (probably) out there — get...   
17           112552                   Joseph Pine: What consumers want   
18           109921  Jinha Lee: Reach into the computer and grab a ...   
19            94321        Clay Shirky: Institutions vs. collaboration   
20           729312  One woman, eight hilarious characters | Sarah ...   
21           287048               How to find a wonderful idea | OK Go   
22           827676  Hans Rosling: Global population growth, box by...   
23            61354   Lian Pin Koh: A drone's-eye view of conservation   
24            33184  Jessica Green: Are we filtering the wrong micr...   
25           110647  Lee Mokobe: A powerful poem about what it feel...   
26            44115  Roger Stein: A bold new way to fund drug research   
27           234990    Why our screens make us less happy | Adam Alter   
28          1289448             Underwater astonishments | David Gallo   
29            77605  How radio telescopes show us unseen galaxies |...   
...             ...                                                ...   
2462         225847                Stuart Brown: Play is more than fun   
2463          30407  Boghuma Kabisen Titanji: Ethical riddles in HI...   
2464         103549              Bruno Maisonnier: Dance, tiny robots!   
2465          50790  Pearl Arredondo: My story, from gangland daugh...   
2466        1084846  Incredible high school musicians from Venezuel...   
2467         477762  On tennis, love and motherhood | Serena Willia...   
2468       13838924     Do schools kill creativity? | Sir Ken Robinson   
2469         150652      Aaron Huey: America's native prisoners of war   
2470         145161  Lessons from the longest study on human develo...   
2471          16645  Rob Harmon: How the market can keep streams fl...   
2472          61998         Emily Pilloton: Teaching design for change   
2473          94867              Ed Boyden: A light switch for neurons   
2474          39603          Alan Siegel: Let's simplify legal jargon!   
2475         349590                            Ze Frank's web playroom   
2476          95681  Seth Priebatsch: Building the game layer on to...   
2477          63575  George Ayittey: Cheetahs vs. Hippos for Africa...   
2478         173724  Marcin Jakubowski: Open-sourced blueprints for...   
2479         223198  Susan Savage-Rumbaugh: The gentle genius of bo...   
2480         112813  Rajesh Rao: Computing a Rosetta Stone for the ...   
2481          81385    Paul Debevec animates a photo-real digital face   
2482         127911  George Monbiot: For more wonder, rewild the world   
2483          96338  How (and why) Russia hacked the US election | ...   
2484      

In [29]:
youtube_data['author3'] = youtube_data['author3'].fillna(youtube_data['author1'])
youtube_data['author3'] = youtube_data['author3'].fillna(youtube_data['author2'])

In [30]:
youtube_data

youtube_views                                      youtube_title  \
0             20695  Inge Missmahl brings peace to the minds of Afg...   
1            129968  Edith Widder: The weird and wonderful world of...   
2             46844     Jacqueline Novogratz on an escape from poverty   
3            281435            Math class needs a makeover | Dan Meyer   
4            192313              Craig Venter unveils "synthetic life"   
5             50511              Danielle de Niese: A flirtatious aria   
6             50557  Philip Howard: Four ways to fix a broken legal...   
7            981094  Why you should define your fears instead of yo...   
8             35683    Dennis vanEngelsdorp: Where have the bees gone?   
9           1751099   Close-up card magic with a twist | Lennart Green   
10           350079            Joshua Klein: The intelligence of crows   
11          1861625           The psychology of evil | Philip Zimbardo   
12           627112              How to make hard choices | Ruth Chang   
13            61947  Dan Barasch: A park underneath the hustle and ...   
14           137256  Jill Tarter: Why the search for alien intellig...   
15            36779              Liz Diller: A giant bubble for debate   
16            42353  Seth Shostak: ET is (probably) out there — get...   
17           112552                   Joseph Pine: What consumers want   
18           109921  Jinha Lee: Reach into the computer and grab a ...   
19            94321        Clay Shirky: Institutions vs. collaboration   
20           729312  One woman, eight hilarious characters | Sarah ...   
21           287048               How to find a wonderful idea | OK Go   
22           827676  Hans Rosling: Global population growth, box by...   
23            61354   Lian Pin Koh: A drone's-eye view of conservation   
24            33184  Jessica Green: Are we filtering the wrong micr...   
25           110647  Lee Mokobe: A powerful poem about what it feel...   
26            44115  Roger Stein: A bold new way to fund drug research   
27           234990    Why our screens make us less happy | Adam Alter   
28          1289448             Underwater astonishments | David Gallo   
29            77605  How radio telescopes show us unseen galaxies |...   
...             ...                                                ...   
2462         225847                Stuart Brown: Play is more than fun   
2463          30407  Boghuma Kabisen Titanji: Ethical riddles in HI...   
2464         103549              Bruno Maisonnier: Dance, tiny robots!   
2465          50790  Pearl Arredondo: My story, from gangland daugh...   
2466        1084846  Incredible high school musicians from Venezuel...   
2467         477762  On tennis, love and motherhood | Serena Willia...   
2468       13838924     Do schools kill creativity? | Sir Ken Robinson   
2469         150652      Aaron Huey: America's native prisoners of war   
2470         145161  Lessons from the longest study on human develo...   
2471          16645  Rob Harmon: How the market can keep streams fl...   
2472          61998         Emily Pilloton: Teaching design for change   
2473          94867              Ed Boyden: A light switch for neurons   
2474          39603          Alan Siegel: Let's simplify legal jargon!   
2475         349590                            Ze Frank's web playroom   
2476          95681  Seth Priebatsch: Building the game layer on to...   
2477          63575  George Ayittey: Cheetahs vs. Hippos for Africa...   
2478         173724  Marcin Jakubowski: Open-sourced blueprints for...   
2479         223198  Susan Savage-Rumbaugh: The gentle genius of bo...   
2480         112813  Rajesh Rao: Computing a Rosetta Stone for the ...   
2481          81385    Paul Debevec animates a photo-real digital face   
2482         127911  George Monbiot: For more wonder, rewild the world   
2483          96338  How (and why) Russia hacked the US election | ...   
2484      

Now by separating the unique column in ted data and youtube data then we can join by authors

In [53]:
 pd.merge(ted_data,youtube_data,left_on='name_slice',right_on='talk_slice')

ted_nof_comments  ted_duration       ted_event     film_dt  \
0                  265           977         TED2006   2/25/2006   
1                  124          1286         TED2006   2/24/2006   
2                  200          1116         TED2006   2/26/2006   
3                  672          1305         TED2006    2/2/2006   
4                  919           992         TED2006   2/24/2006   
5                  900          1262         TED2006   2/25/2006   
6                   99          1057         TED2006   2/23/2006   
7                   43          1000         EG 2007  12/12/2007   
8                  325          1481         TED2006   2/23/2006   
9                  305          1445         TED2004   2/26/2004   
10                 185          1136         TED2004   2/24/2004   
11                 556          1407         TED2006   2/24/2006   
12                 184          1140  TEDGlobal 2005   7/14/2005   
13                 507          1316  TEDGlobal 2005    7/7/2005   
14                  95          1275         TED2004   2/27/2004   
15                 355          1050         TED2004   2/26/2004   
16                 991          1177  TEDGlobal 2005   7/15/2005   
17                 612          1365  TEDGlobal 2005   7/14/2005   
18                  58           773  TEDGlobal 2005    7/1/2005   
19                 494           805         TED2006   2/23/2006   
20                 233          1376         TED2005   2/24/2005   
21                 149          1200         TED2005   2/24/2005   
22                  95           247   TEDSalon 2006  11/30/2006   
23                 499          1001         TED2005    2/2/2005   
24                  91          1141  TEDGlobal 2005   7/14/2005   
25                 111          1385         TED2005   2/25/2005   
26                 161          1645         TED2002    3/1/2002   
27                 220          1021         TED2003   2/27/2003   
28                 114          1211         TED2005    2/2/2005   
29                  49          1340         TED2003    2/2/2003   
...                ...           ...             ...         ...   
1963                81           736         TED2017   4/24/2017   
1964                13           685    TEDxStanford   4/24/2016   
1965                12           530         TED2017   4/24/2017   
1966                20           747    TEDxCrenshaw   10/8/2016   
1967                 9           310         TED2017   4/24/2017   
1968                18           791      TEDxSydney   5/24/2016   
1969                39           817         TED2017   4/24/2017   
1970                17           865    TEDxStanford   4/23/2017   
1971                 3           717     TEDxHamburg    6/8/2016   
1972                45          1156      TEDxExeter   4/24/2015   
1973                12           738         TED2017   4/24/2017   
1974                18           698       TEDxGhent   6/18/2016   
1975                25           993         TED2017   4/24/2017   
1976                 9           663         TED2017   4/24/2017   
1977                 2           725       TEDxSkoll    4/7/2017   
1978                 8           723  TEDGlobal 2017   8/27/2017   
1979                21           865    TEDxMileHigh    7/8/2017   
1980                39           738         TED2017   4/24/2017   
1981                 7           808      TEDxZurich  11/11/2016   
1982                10           934         TED2017   8/27/2017   
1983                33           722         TED@UPS   7/20/2017   
1984                12           957  TEDxManchester   2/19/2017   
1985                 3           795  TEDGlobal 2017   8/27/2017   
1986                 7           894         TED2017   4/24/2017   
1987                10           810  TEDGlobal 2017   8/27/2017   
1988                17           476         TED2017    6/6/2017   
1989                 6           290         TED2017   4/24/20

Youtube data have 2492 rows but the mapped no of columns are 1993. So using outer join to find unmatched rows

In [31]:
data_join_outer = pd.merge(ted_data,youtube_data,left_on='name_slice',right_on='talk_slice',indicator=True,how='outer')

In [32]:
uni_youtube = data_join_outer[data_join_outer['_merge']=='right_only']

In [35]:
uni_youtube.columns

Index(['ted_nof_comments', 'ted_duration', 'ted_event', 'film_dt',
       'ted_avlb_lang', 'main_speaker', 'name', 'num_speaker',
       'ted_published_dt', 'ted_ratings', 'ted_talks_related',
       'speaker_occupation', 'ted_tags', 'ted_title', 'ted_views',
       'name_slice', 'youtube_views', 'youtube_title', 'youtube_tags',
       'youtube_categories', 'youtube_duration', 'youtube_dislikes',
       'youtube_likes', 'youtube_avg_ratings', 'melt1', 'talk1', 'author1',
       'melt2', 'talk2', 'author2', 'talk3', 'talk_slice', 'split', 'author3',
       '_merge'],
      dtype='object')

In [38]:
uni_youtube.drop(['ted_nof_comments', 'ted_duration', 'ted_event',
       'film_dt', 'ted_avlb_lang', 'main_speaker', 'name', 'num_speaker',
        'ted_published_dt', 'ted_ratings', 'ted_talks_related',
       'speaker_occupation', 'ted_tags', 'ted_title', 'name_slice'],inplace=True,axis=1)

C:\Users\Sivakumar\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


In [39]:
uni_ted = data_join_outer[data_join_outer['_merge']=='left_only']

In [40]:
uni_ted.columns

Index(['ted_nof_comments', 'ted_duration', 'ted_event', 'film_dt',
       'ted_avlb_lang', 'main_speaker', 'name', 'num_speaker',
       'ted_published_dt', 'ted_ratings', 'ted_talks_related',
       'speaker_occupation', 'ted_tags', 'ted_title', 'ted_views',
       'name_slice', 'youtube_views', 'youtube_title', 'youtube_tags',
       'youtube_categories', 'youtube_duration', 'youtube_dislikes',
       'youtube_likes', 'youtube_avg_ratings', 'melt1', 'talk1', 'author1',
       'melt2', 'talk2', 'author2', 'talk3', 'talk_slice', 'split', 'author3',
       '_merge'],
      dtype='object')

In [41]:
uni_ted.drop(['youtube_views', 'youtube_title', 'youtube_categories', 'youtube_duration',
       'youtube_dislikes', 'youtube_likes', 'youtube_avg_ratings',
       'melt1', 'talk1', 'author1', 'melt2', 'author2', 'talk2', 'talk3',
       'talk_slice', '_merge'],inplace=True,axis=1)

C:\Users\Sivakumar\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


In [42]:
data_merge = pd.merge(uni_ted,uni_youtube,left_on='main_speaker',right_on='author3')

In [43]:
data_merge

ted_nof_comments  ted_duration               ted_event     film_dt  \
0               593.0        1190.0                 TED2006   2/22/2006   
1               125.0         602.0                 TED2009    2/6/2009   
2                46.0        1198.0                 TED2006   2/23/2006   
3               852.0        1485.0                 TED2006    2/2/2006   
4                79.0        1414.0                 TED2006   2/26/2006   
5                55.0        1538.0                 TED2006   2/26/2006   
6                71.0        1550.0                 TED2006   2/23/2006   
7                71.0        1550.0                 TED2006   2/23/2006   
8                46.0        1261.0  Skoll World Forum 2007    1/1/2007   
9                46.0        1261.0  Skoll World Forum 2007    1/1/2007   
10               88.0         906.0                 TED2006   2/24/2006   
11              163.0        1170.0                 TED2005   2/25/2005   
12               84.0        1201.0          TEDGlobal 2005   7/14/2005   
13              108.0        1114.0                 TED2006   2/23/2006   
14               50.0        1006.0                 TED2006   2/23/2006   
15              117.0        1225.0                 TED2004    2/4/2004   
16               97.0         825.0          TEDGlobal 2005    7/7/2005   
17              213.0        1129.0                 TED2005   2/26/2005   
18               74.0         952.0          TEDGlobal 2005   7/13/2005   
19               43.0        1080.0          TEDGlobal 2005   7/14/2005   
20               67.0        1125.0          TEDGlobal 2005   7/12/2005   
21              186.0        1177.0                 TED2006   2/24/2006   
22               57.0        1083.0                 TED2005   2/24/2005   
23              112.0        1672.0                 TED2005   2/24/2005   
24               58.0        2065.0                 TED2005   2/17/2005   
25               33.0        1609.0                 TED2005   2/24/2005   
26              105.0        1280.0          TEDGlobal 2005   7/14/2005   
27               52.0         848.0                 TED2006   2/23/2006   
28              199.0         198.0                 TED2006   2/23/2006   
29              102.0         192.0                 TED2008    3/3/2008   
..                ...           ...                     ...         ...   
336              94.0         719.0          TEDGlobal 2012   6/27/2012   
337             107.0        1111.0            TEDxWinnipeg   9/13/2012   
338              84.0         535.0           TEDYouth 2011  11/19/2011   
339              91.0         523.0                 TED2013   2/25/2013   
340             225.0        1110.0                 TED2013    3/1/2013   
341             128.0         480.0               TED@Intel    4/7/2013   
342            1000.0         958.0             TEDMED 2013    4/5/2013   
343              85.0         978.0              TEDxAustin    2/9/2013   
344             121.0         476.0            TEDxNijmegen   4/18/2013   
345             218.0         988.0          TEDGlobal 2013   6/14/2013   
346             147.0         645.0          TEDGlobal 2013   6/26/2013   
347              77.0         620.0          TEDxCaFoscariU   6/19/2013   
348             170.0         830.0          TEDGlobal 2013   6/19/2013   
349             178.0         919.0                 TED2013   6/14/2013   
350             177.0         641.0  TEDxHousesOfParliament   6/20/2013   
351             215.0         971.0                 TED2014   3/18/2014   
352             173.0         794.0                 TED2014   3/20/2014   
353              82.0         559.0                 TED2014   3/12/2014   
354             138.0         595.0    TEDxMidAtlantic 2013   10/1/2013   
355              81.0         952.0          TEDGlobal 2013   6/13/2013   
356             335.0        1205.0              TEDxExeter   3/24/2014   
357              56.0        1

In [44]:
data_merge[['talk_slice','name_slice']]

talk_slice                 name_slice
0    hansroslingdebunkingthird  hansroslingthebeststatsyo
1    hansroslingdebunkingthird  hansroslinginsightsonhivi
2    joshuaprinceramusdesignin  joshuaprinceramusbehindth
3    dandennettrespondingtopas  dandennettletsteachreligi
4    cameronsinclairacallforop  cameronsinclairmywishacal
5    jehanenoujaimtedprizewish  jehanenoujaimmywishagloba
6    larrybrilliantthecasefori  larrybrilliantmywishhelpm
7    larrybrillianttedprizewis  larrybrilliantmywishhelpm
8    larrybrilliantthecasefori  larrybrilliantthecaseforo
9    larrybrillianttedprizewis  larrybrilliantthecaseforo
10   amysmithsimpledesignsthat  amysmithsimpledesignstosa
11   rosslovegrovethepowerandb  rosslovegroveorganicdesig
12   jimmywaleshowaragtagbandc  jimmywalesthebirthofwikip
13   richardbaraniukonopensour  richardbaraniukthebirthof
14   menatrotthowblogsarebuild  menatrottmeetthefounderof
15   eveenslersecurityandinsec  eveenslerhappinessinbodya
16   eveenslersecurityandinsec  eveenslerwhatsecuritymean
17   evavertesmydreamaboutthef  evavertesmeetthefutureofc
18   iqbalquadirthepowerofthem  iqbalquadirhowmobilephone
19   sasavucinicwhyafreepressi  sasavucinicwhyweshouldinv
20   ashrafghanihowtofixbroken  ashrafghanihowtorebuildab
21   burtrutanentrepreneursare  burtrutantherealfutureofs
22   bensaundersthreethingstok  bensaunderswhydidiskitoth
23         bonoactionforafrica  bonomywishthreeactionsfor
24   edwardburtynskymanufactur  edwardburtynskymywishmanu
25   robertfischelltedprizewis  robertfischellmywishthree
26   peterdonnellyhowstatsfool  peterdonnellyhowjuriesare
27   petergabrielfightinginjus  petergabrielfightinjustic
28   deanornishtheworldskiller  deanornishthekillerameric
29   deanornishtheworldskiller  deanornishyourgenesarenot
..                         ...                        ...
336  andrewblumwhatistheintern  andrewblumdiscoverthephys
337  jeffhancock3typesofdigita  jeffhancockthefutureoflyi
338  lemonandersenperformsplea  lemonandersenpleasedontta
339  jessicagreengoodgermsmake  jessicagreenwerecoveredin
340  joshuapragerinsearchforth  joshuapragerinsearchofthe
341  mariabezaitisthesurprisin  mariabezaitiswhyweneedstr
342  peterattiawhatifwerewrong  peterattiaistheobesitycri
343  joelselanikiothesurprisin  joelselanikiothebigdatare
344  petervanmanenhowcanformul  petervanmanenbetterbabyca
345  michaelporterwhybusinessc  michaelporterthecaseforle
346    tritaparsiiranandisrael  tritaparsiiranandisraelpe
347  frederickaplanhowibuiltan  frederickaplanhowtobuilda
348            markkendalldemo  markkendalldemoaneedlefre
349  anantagarwalwhymassivelyo  anantagarwalwhymassiveope
350  davidputtnamwhathappenswh  davidputtnamdoesthemediah
351        charmiangoochmywish  charmiangoochmywishtolaun
352  edyongsuicidalwaspszombie  edyongzombieroachesandoth
353  delharveythestrangenessof  delharveyprotectingtwitte
354  jennifergolbeckthecurlyfr  jennifergolbeckyoursocial
355  urialonwhytrulyinnovative  urialonwhysciencedemandsa
356  karimabennounethesideofte  karimabennounewhenpeopleo
357  colingrantthesonofadiffic  colingranthowourstoriescr
358  nancyfrateswhymyfamilysta  nancyfratesmeetthemomwhos
359  rainerstrackthesurprising  rainerstracktheworkforcec
360  josemiguelsokoloffhowweus  josemiguelsokoloffhowchri
361  zeyneptufekcihowtheintern  zeyneptufekcionlinesocial
362  ricardosemlerradicalwisdo  ricardosemlerhowtorunacom
363  shimpeitakahashiplaythisg  shimpeitakahashiplaythisw
364  taldaninowecanusebacteria  taldaninoprogrammingbacte
365  margaretheffernanwhyitsti  margaretheffernanforgetth

[366 rows x 2 columns]

While split and merging there may be chances of more than one talk for a speaker, so filtering of one talk by one speaker

In [45]:
data_merge.groupby('author3_y')['name'].count().sort_values(ascending=False)

author3_y
Steven Johnson          4
Jay Walker              4
Murray Gell-Mann        4
Craig Venter            4
Golan Levin             4
Nathan Myhrvold         4
Frank Gehry             4
David Hoffman           4
Raul Midon              4
Liz Diller              4
Ngozi Okonjo-Iweala     4
Jacqueline Novogratz    4
David Pogue             4
Stewart Brand           4
Larry Brilliant         4
Robert Full             4
Dean Kamen              4
Paul Rothemund          4
E.O. Wilson             4
Chris Anderson          3
Paul Collier            2
David Gallo             2
John Lloyd              2
Hans Rosling            2
Lewis Pugh              2
Eve Ensler              2
Kary Mullis             2
Marco Tempest           2
Nicholas Negroponte     2
Eddi Reader             2
                       ..
Peter Attia             1
Pete Alcorn             1
Penelope Boston         1
Pawan Sinha             1
Paul Sereno             1
Paul Moller             1
Paul MacCready          1
Pa

In [54]:
single_talk = data_merge.groupby('author3_y').filter(lambda g:g['name'].count()==1)

In [55]:
single_talk

ted_nof_comments  ted_duration               ted_event     film_dt  \
2                46.0        1198.0                 TED2006   2/23/2006   
3               852.0        1485.0                 TED2006    2/2/2006   
4                79.0        1414.0                 TED2006   2/26/2006   
5                55.0        1538.0                 TED2006   2/26/2006   
10               88.0         906.0                 TED2006   2/24/2006   
11              163.0        1170.0                 TED2005   2/25/2005   
12               84.0        1201.0          TEDGlobal 2005   7/14/2005   
13              108.0        1114.0                 TED2006   2/23/2006   
14               50.0        1006.0                 TED2006   2/23/2006   
17              213.0        1129.0                 TED2005   2/26/2005   
18               74.0         952.0          TEDGlobal 2005   7/13/2005   
19               43.0        1080.0          TEDGlobal 2005   7/14/2005   
20               67.0        1125.0          TEDGlobal 2005   7/12/2005   
21              186.0        1177.0                 TED2006   2/24/2006   
22               57.0        1083.0                 TED2005   2/24/2005   
23              112.0        1672.0                 TED2005   2/24/2005   
24               58.0        2065.0                 TED2005   2/17/2005   
25               33.0        1609.0                 TED2005   2/24/2005   
26              105.0        1280.0          TEDGlobal 2005   7/14/2005   
27               52.0         848.0                 TED2006   2/23/2006   
30               85.0         843.0          TEDGlobal 2005   7/14/2005   
31              371.0        1321.0                 TED2003    2/3/2003   
32               59.0        1115.0                 TED2006   2/23/2006   
33              130.0        1151.0                 TED2006    2/2/2006   
38               28.0         869.0                 TED2006   2/23/2006   
39               64.0        1038.0                 TED2006    2/2/2006   
40               97.0        1447.0                 TED2007    3/8/2007   
51               50.0        1240.0                 TED2005   2/25/2005   
52               78.0        1204.0                 TED2002    2/2/2002   
53              131.0        1012.0          TEDGlobal 2005   7/14/2005   
..                ...           ...                     ...         ...   
336              94.0         719.0          TEDGlobal 2012   6/27/2012   
337             107.0        1111.0            TEDxWinnipeg   9/13/2012   
338              84.0         535.0           TEDYouth 2011  11/19/2011   
339              91.0         523.0                 TED2013   2/25/2013   
340             225.0        1110.0                 TED2013    3/1/2013   
341             128.0         480.0               TED@Intel    4/7/2013   
342            1000.0         958.0             TEDMED 2013    4/5/2013   
343              85.0         978.0              TEDxAustin    2/9/2013   
344             121.0         476.0            TEDxNijmegen   4/18/2013   
345             218.0         988.0          TEDGlobal 2013   6/14/2013   
346             147.0         645.0          TEDGlobal 2013   6/26/2013   
347              77.0         620.0          TEDxCaFoscariU   6/19/2013   
348             170.0         830.0          TEDGlobal 2013   6/19/2013   
349             178.0         919.0                 TED2013   6/14/2013   
350             177.0         641.0  TEDxHousesOfParliament   6/20/2013   
351             215.0         971.0                 TED2014   3/18/2014   
352             173.0         794.0                 TED2014   3/20/2014   
353              82.0         559.0                 TED2014   3/12/2014   
354             138.0         595.0    TEDxMidAtlantic 2013   10/1/2013   
355              81.0         952.0          TEDGlobal 2013   6/13/2013   
356             335.0        1205.0              TEDxExeter   3/24/2014   
357              56.0        1

    So now 1993 + 259 we can able to match upto 2252 rows

In [ ]:
youtube_data[['youtube_title','youtube_views']].sort_values(by='youtube_views',ascending = False).head(10)

In [ ]:
ted_data[['name','views']].sort_values(by='views',ascending = False).head(10)